hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AzureBlobBlockManager.java

```diff
@@ -0,0 +1,256 @@
+   * @throws IOException if an I/O error occurs
+   */
+  @Override
+  protected synchronized AbfsBlock createBlockInternal(final long position)
+      throws IOException {
+    if (activeBlock == null) {
+      blockCount++;
+      activeBlock = new AbfsBlobBlock(abfsOutputStream, position);
+    }
+    return activeBlock;
+  }
+
+  /**
+   * Returns block id's which are committed for the blob.
+   *
+   * @param tracingContext Tracing context object.
+   * @return list of committed block id's.
+   * @throws AzureBlobFileSystemException if an error occurs
+   */
+  private List<String> getBlockList(TracingContext tracingContext)
+      throws AzureBlobFileSystemException {
+    List<String> committedBlockIdList;
+    AbfsBlobClient blobClient = abfsOutputStream.getClientHandler().getBlobClient();
+    final AbfsRestOperation op = blobClient
+        .getBlockList(abfsOutputStream.getPath(), tracingContext);
+    committedBlockIdList = op.getResult().getBlockIdList();
+    return committedBlockIdList;
+  }
+
+  /**
+   * Adds the block entry to the map with status NEW.
+   *
+   * @param block the block to track
+   */
+  protected void trackBlockWithData(AbfsBlock block) {
+    lock.lock();
+    try {
+      blockStatusMap.put(block.getBlockId(), AbfsBlockStatus.NEW);
+      orderedBlockList.add(block.getBlockId());
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  /**
+   * Updates the status of the specified block.
+   *
+   * @param block the block to update
+   * @param status the new status
+   * @throws IOException if an I/O error occurs
+   */
+  protected void updateBlockStatus(AbfsBlock block, AbfsBlockStatus status)
+      throws IOException {
+    String key = block.getBlockId();
+    lock.lock();
+    try {
+      if (!getBlockStatusMap().containsKey(key)) {
+        throw new IOException("Block is missing with blockId " + key
+            + " for offset " + block.getOffset()
+            + " for path" + abfsOutputStream.getPath()
+            + " with streamId " + abfsOutputStream.getStreamID());
+      } else {
+        blockStatusMap.put(key, status);
+      }
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  /**
+   * Prepares the list of blocks to commit.
+   *
+   * @param offset the offset
+   * @return the number of blocks to commit
+   * @throws IOException if an I/O error occurs
+   */
+  protected int prepareListToCommit(long offset) throws IOException {
+    // Adds all the committed blocks if available to the list of blocks to be added in putBlockList.
+    blockIdList.addAll(committedBlockEntries);
+    String failedBlockId;
+    AbfsBlockStatus success = AbfsBlockStatus.SUCCESS;
+
+    // No network calls needed for empty map.
+    if (blockStatusMap.isEmpty()) {
+      return 0;
+    }
+
+    int mapEntry = 0;
+    // If any of the entry in the map doesn't have the status of SUCCESS, fail the flush.
+    for (Map.Entry<String, AbfsBlockStatus> entry : getBlockStatusMap().entrySet()) {
```

Comment:
I believe that this block is to verify if the AbfsOutputStream has correctly added blockIds in correct sequence. Is there any more reason?

If its to verify, I would be more inclined towards removing this block  -> this would save computation at flush + computation to set blockStatusMap. And as far as the verification is concerned, I would say our integ tests should give that confidence (plus if this is breaking, this will break at large scale).

Reply:
If we remove this block, I would suggest further to remove blockStatusMap totally. `orderedBlockList` is the important data-structure for us.

Reply:
Appreciate the suggestion here but would like to keep this code block as it is. Computation cost is not much here.

Reply:
I believe that this block is to verify if the AbfsOutputStream has correctly added blockIds in correct sequence. Is there any more reason? -> this is not just for sequence, it validates that all appends are successful before flushing

Reply:
This code would be firing at `flushWrittenBytesToServiceInternal` which has two callers:
1. flushWrittenBytesToServiceAsync
2. flushWrittenBytesToService
In both the cases, flushWrittenBytesToServiceInternal can be only called when the parallel threads are completely done sucessfully.

So, technically this flush api on server would be getting called only once the writing is successful on all the threads. As per code, the data-consistency for write / flush is maintained.